music_assistant/providers/audioscrobbler/__init__.py

```diff
@@ -0,0 +1,271 @@
+from music_assistant.helpers.auth import AuthenticationHelper
+from music_assistant.models import ProviderInstanceType
+from music_assistant.models.plugin import PluginProvider
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    return ScrobblerProvider(mass, manifest, config)
+
+
+class ScrobblerProvider(PluginProvider):
+    """Plugin provider to support scrobbling of tracks."""
+
+    _network: pylast._Network = None
+
+    def __init__(
+        self, mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+    ) -> None:
+        """Initialize AudioScrobbler."""
+        super().__init__(mass, manifest, config)
+        if self.logger.level == logging.DEBUG:
+            pylast.logger.setLevel(logging.DEBUG)
+
+    def _get_network_config(self) -> dict[str, ConfigValueType]:
+        return {
+            CONF_API_KEY: self.config.get_value(CONF_API_KEY),
+            CONF_API_SECRET: self.config.get_value(CONF_API_SECRET),
+            CONF_PROVIDER: self.config.get_value(CONF_PROVIDER),
+            CONF_USERNAME: self.config.get_value(CONF_USERNAME),
+            CONF_SESSION_KEY: self.config.get_value(CONF_SESSION_KEY),
+        }
+
+    async def loaded_in_mass(self) -> None:
+        """Call after the provider has been loaded."""
+        await super().loaded_in_mass()
+
+        if self.config.get_value(CONF_SESSION_KEY):
+            self._network = _get_network(self._get_network_config())
+
+        # could be interesting for a "scrobble after X seconds" feature
+        # self.mass.subscribe(self._on_mass_queue_time_updated, EventType.QUEUE_TIME_UPDATED)
+        self.mass.subscribe(self._on_mass_queue_updated, EventType.QUEUE_UPDATED)
+        self.mass.subscribe(self._on_mass_media_item_played, EventType.MEDIA_ITEM_PLAYED)
+
+    async def _on_mass_queue_updated(self, event: MassEvent) -> None:
+        """Player has updated, update nowPlaying."""
+        if self._network is None:
+            return
+
+        queue: PlayerQueue = event.data
+        if queue.state == PlayerState.PLAYING and queue.current_item.media_type == MediaType.TRACK:
+            track: Track = queue.current_item.media_item
+            try:
+                self._network.update_now_playing(
+                    artist=track.artist_str,
+                    title=track.name,
+                    mbid=track.mbid,
+                    album=track.album.name if track.album else None,
+                )
+            except Exception as err:
+                self.logger.exception(err)
+
+    async def _on_mass_media_item_played(self, event: MassEvent) -> None:
+        """Media item has finished playing, we'll scrobble the track."""
+        if self._network is None:
+            return
+
+        item = await self.mass.music.get_item_by_uri(event.object_id)
```

Comment:
event.data also contains a media_item attribute so you wont have to fetch it , saves a potential api call

Reply:
doh, no idea how I missed that, I'll update that, thanks!

Reply:
I figured out why I missed that. Because it's not true :P `data["media_item']` contains the reference as well, not the item. I played around a bit with passing the actual item, but I got serialization errors.

Reply:
I updated the object with the info present. We must prevent fetching full item details where/if we can to save api calls